analyses/malawi/dryspells_trigger/07_ecmwf_monthly_skill_general.md

```diff
@@ -41,164 +37,372 @@ reload(processing)
+
+
+Questions: 
+- better use mean or median to compute the difference between observation and forecast? As they give quite different results 
+- Does using the CI make sense or better use percentiles? Due to the large number of values the CI becomes small (which can be a good thing), but at the same time the std is quite large
+
+```python
+def calc_diff_stats(observations,forecast):
+    diff=forecast-observations
+    diff_mean=diff.mean()
+    diff_median=diff.median()
+    diff_std=diff.std()
+    diff_count=diff.count()
+    ci95_hi = diff_mean + 1.95*diff_std/math.sqrt(diff_count)
+    ci95_lo = diff_mean - 1.95*diff_std/math.sqrt(diff_count)
+    return diff_median, diff_mean, diff_std, ci95_hi, ci95_lo
+
+def compute_diff_cats(df,obs_col,for_col,threshold_list=None,adm_list=None,month_list=None):
+    leadtimes=df.leadtime.sort_values().unique()
+    df_diff=pd.DataFrame(leadtimes,columns=["leadtime"]).set_index('leadtime')
+    for ilt, leadtime in enumerate(leadtimes):
+        df_lt=df[df.leadtime==leadtime]
+        med,me,std,ci_hi,ci_lo=calc_diff_stats(df_lt[obs_col],df_lt[for_col])
+        df_diff.loc[leadtime,"all_med"]=med
+        df_diff.loc[leadtime,"all_mean"]=me
+        df_diff.loc[leadtime,"all_ci_lo"]=ci_lo
+        df_diff.loc[leadtime,"all_ci_hi"]=ci_hi
+        if threshold_list:
+            for thresh in threshold_list:
+                df_lt_thresh=df_lt[df_lt[obs_col]<=thresh]
+                med,me,std,ci_hi,ci_lo = calc_diff_stats(df_lt_thresh[obs_col],df_lt_thresh[for_col])
+                df_diff.loc[leadtime,f"thresh_{thresh}_med"]=med
+        if month_list:
+            for m in month_list:
+                month_str="".join([calendar.month_abbr[i].lower() for i in m])
+                df_lt_m=df_lt[df_lt.time.dt.month.isin(m)]
+                med,me,std,ci_hi,ci_lo = calc_diff_stats(df_lt_m[obs_col],df_lt_m[for_col])
+                df_diff.loc[leadtime,f"{month_str}_med"]=med
+                df_diff.loc[leadtime,f"{month_str}_ci_lo"]=ci_lo
+                df_diff.loc[leadtime,f"{month_str}_ci_hi"]=ci_hi
+                
+    return df_diff
+```
+
+```python
+df_diff=compute_diff_cats(df_forobs,"precip_obs","precip_for",threshold_list=[210,180,170],month_list=[[m] for m in range(1,13)])
+```
+
+```python
+fig, ax = plt.subplots()
+ax.plot(df_diff.index,df_diff["all_mean"],label="mean")
+ax.fill_between(df_diff.index, df_diff.all_ci_lo, df_diff.all_ci_hi,alpha=0.2,label="confidence interval")
+ax.plot(df_diff.index, df_diff["all_med"],label="median")
+ax.set_xlabel("leadtime [months]")
+ax.set_ylabel("median bias")
+ax.grid()
+ax.legend(bbox_to_anchor=(1.05, 1));
+```
+
+When looking per month though, we can see that the average bias clearly differs per month. The months around the rainy season have the highest bias (nov, dec, jan, feb). This can either be because values are generally larger, or because the forecast has less skill. 
```

Comment:
I'm not sure I understand exactly what's plotted here. When I took a quick peak at a subset of the data (e.g. `df_forobs[(df_forobs['month']==1) & (df_forobs['leadtime']==1)]['diff_forobs']`) the mean / median are closer to 40 / 50.

Reply:
That is super strange. Cause that is exactly what the plot shows, but when I do `df_forobs[(df_forobs['month']==1) & (df_forobs['leadtime']==1)]['diff_forobs'].median()` , I do get the values as shown in the plot (in this case 26.64). Could you try again?

Reply:
I plotted again and it looks much more reasonable: 
![image](https://user-images.githubusercontent.com/6585693/127856505-a95a7a07-68d6-4ca2-ac7c-9f29aa1879ae.png)